File: typescript/min_heap.ts
class MinHeap {
    length: number;
    array: number[];

    constructor() {
        this.length = 0;
        this.array = [];
    }

    insert(item: number): void {
        this.array[this.length] = item;
        this.length += 1;
        this.upHeap(this.length - 1);
    }

    delete(): number | undefined {
        if (this.length === 0) {
            return undefined;
        }

        const out = this.array[0];
        this.length--;
        this.array[0] = this.array[this.length];
        this.array.splice(this.length, 1);
        this.downHeap(0);
        return out;
    }

    private downHeap(idx: number): void {
        const leftIdx = this.leftIdx(idx);
        const rightIdx = this.rightIdx(idx);
        let smallest = idx;

        if (leftIdx < this.length && this.array[leftIdx] < this.array[idx]) {
            smallest = leftIdx;
        }

        if (rightIdx < this.length && this.array[rightIdx] < this.array[smallest]) {
            smallest = rightIdx;
        }

        if (smallest !== idx) {
            const temp = this.array[idx];
            this.array[idx] = this.array[smallest];
            this.array[smallest] = temp;
            this.downHeap(smallest);
        }

    }

    private upHeap(idx: number): void {
        if (this.length < 2) {
            return;
        }

        const parentIdx = this.parentIdx(idx);
        const parentVal = this.array[parentIdx];
        const currentVal = this.array[idx];

        if (currentVal < parentVal) {
            this.array[idx] = parentVal;
            this.array[parentIdx] = currentVal;
            this.upHeap(parentIdx);
        }
    }

    private parentIdx(idx: number): number {
        return Math.floor((idx - 1) / 2);
    }

    private leftIdx(idx: number): number {
        return idx * 2 + 1;
    }

    private rightIdx(idx: number): number {
        return idx * 2 + 2;
    }
}

// test
const heap = new MinHeap();

// add
console.log(heap.array);
heap.insert(9);
console.log(heap.array);
heap.insert(4);
console.log(heap.array);
heap.insert(3);
console.log(heap.array);
heap.insert(1);
console.log(heap.array);
heap.insert(7);
console.log(heap.array);

// delete
heap.delete();
console.log(heap.array);
heap.delete();
console.log(heap.array);
heap.delete();
console.log(heap.array);
heap.delete();
console.log(heap.array);
heap.delete();
console.log(heap.array);

File: python/min_heap.py
class MinHeap:
    def __init__(self) -> None:
        self.array: list[int] = []

    def length(self) -> int:
        return len(self.array)

    def insert(self, item: int) -> None:
        self.array.append(item)
        self.up_heap(self.length() - 1)

    def delete(self) -> int | None:
        if self.length() == 0:
            return None

        if self.length() - 1 == 0:
            return self.array.pop()

        item = self.array[0]
        self.array[0] = self.array.pop()
        self.down_heap(0)
        return item

    def up_heap(self, idx: int) -> None:
        if idx < 1:
            return

        parent_idx = self.parent_idx(idx)
        parent_val = self.array[parent_idx]

        if parent_val > self.array[idx]:
            self.array[parent_idx] = self.array[idx]
            self.array[idx] = parent_val
            self.up_heap(parent_idx)

    def down_heap(self, idx: int) -> None:
        left_idx = self.left_idx(idx)
        right_idx = self.right_idx(idx)
        smallest = idx

        if left_idx < self.length() and self.array[left_idx] < self.array[idx]:
            smallest = left_idx

        if right_idx < self.length() and self.array[right_idx] < self.array[smallest]:
            smallest = right_idx

        if smallest != idx:
            self.array[smallest], self.array[idx] = self.array[idx], self.array[smallest]
            self.down_heap(smallest)

    # formula to get parent index
    def parent_idx(self, idx: int) -> int:
        return (idx - 1) // 2

    # formula to get the left child index
    def left_idx(self, idx: int) -> int:
        return idx * 2 + 1

    def right_idx(self, idx: int) -> int:
        return idx * 2 + 2


# test
def test():
    min_heap = MinHeap()
    items = [9, 7, 4, 1, 5, 4, 2]

    print(min_heap.array)
    for i in items:
        min_heap.insert(i)
        print(min_heap.array)

    for _ in items:
        min_heap.delete()
        print(min_heap.array)


test()
